rtos/TARGET_CORTEX/mbed_rtx_idle.cpp

```diff
@@ -0,0 +1,230 @@
+        // to trigger software interrupts on each tick. The period does
+        // not matter since it will never start counting.
+        SysTick_Setup(16);
+#endif
+    };
+
+    /**
+     * Schedule an os tick to fire
+     *
+     * @param delta Tick to fire at relative to current tick
+     */
+    void schedule_tick(uint32_t delta=1) {
+        insert_absolute(_start_time + (_tick + delta) * 1000000 /  OS_TICK_FREQ);
+    }
+
+
+    /**
+     * Prevent any scheduled ticks from triggering
+     */
+    void cancel_tick() {
+        remove();
+    }
+
+    /**
+     * Get the current tick count
+     *
+     * @return The number of ticks since boot. This should match RTX's tick count
+     */
+    uint32_t get_tick() {
+        return _tick & 0xFFFFFFFF;
+    }
+
+    /**
+     * Update the internal tick count
+     *
+     * @return The number of ticks incremented
+     */
+    uint32_t update_tick() {
+        uint64_t new_tick = ticker_read_us(_ticker_data) * OS_TICK_FREQ / 1000000;
+        if (new_tick > _tick) {
+            // Don't update to the current tick. Instead, update to the
+            // previous tick and let the SysTick handler increment it
+            // to the current value. This allows scheduling restart
+            // successfully after the OS is resumed.
+            new_tick--;
+        }
+        uint32_t elapsed_ticks = new_tick - _tick;
+        _tick = new_tick;
+        return elapsed_ticks;
+    }
+
+    /**
+     * Get the time
+     *
+     * @return Current time in microseconds
+     */
+    us_timestamp_t get_time() {
+        return ticker_read_us(_ticker_data);
+    }
+
+    ~RtosTimer() {
+
+    };
+
+protected:
+
+    void handler() {
+#if (defined(NO_SYSTICK))
+        NVIC_SetPendingIRQ(mbed_get_m0_tick_irqn());
+#else
+        SCB->ICSR = SCB_ICSR_PENDSTSET_Msk;
+#endif
+        _tick++;
+    }
+
+    us_timestamp_t _start_time;
+    uint64_t _tick;
+};
+
+static RtosTimer *os_timer;
+static uint64_t os_timer_data[sizeof(RtosTimer) / 8];
+
+/// Setup System Timer.
+int32_t osRtxSysTimerSetup (void)
+{
+    // Do not use SingletonPtr since this relies on the RTOS
+    if (NULL == os_timer) {
+        os_timer = new (os_timer_data) RtosTimer();
+    }
+    return -1;
+}
+
+/// Enable System Timer.
+void osRtxSysTimerEnable (void)
+{
+    // set to fire interrupt on next tick
+    os_timer->schedule_tick();
+}
+
+/// Disable System Timer.
+void osRtxSysTimerDisable (void)
+{
+    os_timer->cancel_tick();
+}
+
+/// Acknowledge System Timer IRQ.
+void osRtxSysTimerAckIRQ (void)
+{
+    os_timer->schedule_tick();
+}
+
+/// Get System Timer count.
+uint32_t osRtxSysTimerGetCount (void)
+{
+    return os_timer->get_time() & 0xFFFFFFFF;
+}
+
+static void default_idle_hook(void)
+{
+    uint32_t elapsed_ticks = 0;
+
+    core_util_critical_section_enter();
+    uint32_t ticks_to_sleep = svcRtxKernelSuspend();
+    MBED_ASSERT(os_timer->get_tick() == svcRtxKernelGetTickCount());
```

Comment:
breaks debugger

Reply:
@saedelman Can you provide more details, or create an issue to reference this line there so we can review?

Reply:
Using gdb and single stepping through the RTX code causes the assert to be triggered. This may be related to the "Gotcha" related to in https://github.com/nuket/mbed-memory-status although I am using SEGGER RTT debug output not serial output:

"On mbed 5.5 and up, there may be a Heisenbug when calling print_thread_info() inside of osKernelLock()!

This error sometimes appears on the serial console shortly after chip startup, but not always: mbed assertation failed: os_timer->get_tick() == svcRtxKernelGetTickCount(), file: .\mbed-os\rtos\TARGET_CORTEX\mbed_rtx_idle.c

The RTOS seems to be asserting an idle constraint violation due to the slowness of sending data through the serial port, but it does not happen consistently."